### PR TITLE
Introduce blocking users from commenting

### PR DESCRIPTION
**Pull request by eduardoj**
These are two screenshots of the user interface after the changes of this pull request are applied:

Screenshot from 2024-02-05 15-08-29
Screenshot from 2024-02-05 15-08-57

### CODE REVIEW

File: src/api/app/policies/comment_policy.rb [hunk 1]
```diff
@@ -5,7 +5,8 @@ def initialize(user, record)
 
   def create?
     return false if user.blank? || user.is_nobody?
-    return true if maintainer? || user.is_admin? || user.is_moderator? || user.is_staff?
+    return true if maintainer? || important_user?
+    return false if Flipper.enabled?(:content_moderation, user) && user.blocked_from_commenting
```
[the remark this answers is not in the document]
Comment (eduardoj) [reply src/api/app/policies/comment_policy.rb:9]: Then, what do we make this feature dependent on, instead?

File: src/api/app/controllers/webui/users_controller.rb [hunk 2]
```diff
@@ -68,7 +68,9 @@ def create
   end
 
   def update
-    if !User.admin_session? && (User.session! != @displayed_user || !@configuration.accounts_editable?(@displayed_user))
+    if !User.admin_session? &&
+       (User.session! != @displayed_user || !@configuration.accounts_editable?(@displayed_user)) &&
+       !(User.session.is_moderator? && params[:user][:blocked_from_commenting].present?)
```
Comment (danidoni) [inline src/api/app/controllers/webui/users_controller.rb:73]: Shouldn't this be handled in a policy?

File: src/api/app/policies/comment_policy.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (eduardoj) [reply src/api/app/policies/comment_policy.rb:9]: Or do we want to introduce this "Blocking users from commenting" feature independent from the feature flag?

[the remark this answers is not in the document]
Comment (eduardoj) [reply src/api/app/policies/comment_policy.rb]: I removed the `Flipper.enabled?` check. Please take a look again.

File: src/api/app/controllers/webui/users_controller.rb [hunk 3]
```diff
@@ -185,6 +187,7 @@ def assign_common_user_attributes
     @displayed_user.assign_attributes(params[:user].slice(:biography, :color_theme).permit!)
     @displayed_user.assign_attributes(params[:user].slice(:realname, :email).permit!) unless @account_edit_link
     @displayed_user.toggle(:in_beta) if params[:user][:in_beta]
+    @displayed_user.toggle(:blocked_from_commenting) if params[:user][:blocked_from_commenting]
```
[the remark this answers is not in the document]
Comment (eduardoj) [reply src/api/app/controllers/webui/users_controller.rb:190]: I used a `button_to`, as you suggested below.

File: src/api/app/views/webui/user/_basic_info.html.haml [hunk 4]
```diff
@@ -53,8 +53,16 @@
...
 
-- if Flipper.enabled?(:content_moderation, User.possibly_nobody)
-  = render partial: 'webui/shared/report_modal'
+  - if Flipper.enabled?(:content_moderation, User.possibly_nobody)
+    = render partial: 'webui/shared/report_modal'
+    - if User.possibly_nobody.is_admin? || User.possibly_nobody.is_moderator?
+      .mt-2
+        = form_with(url: user_path(user), method: :patch) do |_form|
+          .form-check.form-switch
+            = hidden_field_tag('user[blocked_from_commenting]', false, id: nil)
```
[the remark this answers is not in the document]
Comment (eduardoj) [reply src/api/app/views/webui/user/_basic_info.html.haml:62]: Using a button simplified the things a little bit. It was not needed to use a form nor a hidden field. Please take a look again.

File: src/api/app/controllers/webui/users_controller.rb (hunk 2 above)
Comment (eduardoj) [reply src/api/app/controllers/webui/users_controller.rb:73]: It sounds like a pundit policy, yes.

There is code for updating in the user policy: https://github.com/openSUSE/open-build-service/blob/89a89730ed1770996ff86e911374a1dbdb92b790/src/api/app/policies/user_policy.rb#L2-L4

This code seems to be not in sync with the update actions with the API and Webui controllers for a while. I see that refactoring the update actions to use the same update policy is a refactoring opportunity, but I don't see it necessary belonging to this pull request. Do you mind leaving this refactoring outside this pull request?

Comment (danidoni) [reply src/api/app/controllers/webui/users_controller.rb:73]: > Do you mind leaving this refactoring outside this pull request?

I don't mind :+1:

**Review by danidoni — APPROVED** [review]

**Pull request merged**